### ISSUE TITLE
binary-compat: Incompatibility between `struct stat` structure in `nolibc` and in Linux

### ISSUE THREAD

**Issue by razvand** [unikraft/unikraft#498]
### Describe the bug

`struct stat` has a different memory layout (size and offsets of structure members) in `nolibc` and in Linux. This directly binary executables built on Linux that are aimed to run under binary compatibility mode.

For example, running [this sample program](https://github.com/unikraft/static-pie-apps/tree/master/small/stat) under Linux gives the output:

```
 * st_dev:      value:      66309, size:  8, offset:  0
 * st_ino:      value:     935356, size:  8, offset:  8
 * st_mode:     value:      33204, size:  4, offset: 24
 * st_nlink:    value:          1, size:  8, offset: 16
 * st_uid:      value:       1000, size:  4, offset: 28
 * st_gid:      value:       1000, size:  4, offset: 32
 * st_rdev:     value:          0, size:  8, offset: 40
 * st_size:     value:          0, size:  8, offset: 48
 * st_blksize:  value:       4096, size:  8, offset: 56
 * st_blocks:   value:          0, size:  8, offset: 64
 * st_atim:     value:          0, size: 16, offset: 72
 * st_mtim:     value:          0, size: 16, offset: 88
 * st_ctim:     value:          0, size: 16, offset: 104

total size: 144
```

Whereas running it under Unikraft gives:

```
 * st_dev:      value:          0, size:  2, offset:  0
 * st_ino:      value:      14461, size:  2, offset:  2
 * st_mode:     value:      33204, size:  4, offset:  4
 * st_nlink:    value:          0, size:  2, offset:  8
 * st_uid:      value:          0, size:  2, offset: 10
 * st_gid:      value:          0, size:  2, offset: 12
 * st_rdev:     value:          0, size:  2, offset: 14
 * st_size:     value:          0, size:  8, offset: 16
 * st_blksize:  value:        512, size:  8, offset: 72
 * st_blocks:   value:          0, size:  8, offset: 80
 * st_atim:     value:          0, size: 16, offset: 24
 * st_mtim:     value:          0, size: 16, offset: 40
 * st_ctim:     value:          0, size: 16, offset: 56

total size: 88
```

There is a mismatch between the size and offset of structure members, and this affects the running of Linux binary executables in binary compatibility mode on top of Unikraft.

`struct stat` is currently defined in [`nolibc`](https://github.com/unikraft/unikraft/blob/staging/lib/nolibc/include/sys/stat.h#L23) in Unikraft.

### Steps to reproduce

Build and the [sample program](https://github.com/unikraft/static-pie-apps/tree/master/small/stat) in Linux and with Unikraft.

For Unikraft, [newlib](sudo qemu-system-x86_64 -enable-kvm -cpu host -m 2G -nographic -nodefaults -display none -serial stdio -device isa-debug-exit -fsdev local,security_model=passthrough,id=hvirtio0,path=rootfs -device virtio-9p-pci,fsdev=hvirtio0,mount_tag=rootfs -kernel build/app-fs-ops_kvm-x86_64) must be added as a dependency and `9pfs` support must be added. Run the resulting image with a command such as:

```
sudo qemu-system-x86_64 -enable-kvm -cpu host -m 2G -nographic -nodefaults -display none -serial stdio -device isa-debug-exit -fsdev local,security_model=passthrough,id=hvirtio0,path=rootfs -device virtio-9p-pci,fsdev=hvirtio0,mount_tag=rootfs -kernel build/app-fs-ops_kvm-x86_64
```

Replace the `path=rootfs` with the location of the root directory to be mounted via `9pfs` in Unikraft.

### Expected behavior

`struct stat` should have the same layout under Unikraft and under Linux. This will allow Linux binary executables to run under Unikraft.

### Which architectures were you using or does this bug affect?

x86_64

### Which platforms were you using or does this bug affect?

kvm

**Comment by razvand** [comment]
I have a solution for this. I will submit a PR.

**Comment by razvand** [comment]
Solved by #675
